routing/state.go

```diff
@@ -0,0 +1,161 @@
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/peer"
+	pb "github.com/libp2p/go-libp2p-core/routing/pb"
+	ma "github.com/multiformats/go-multiaddr"
+	"time"
+)
+
+// The domain string used for routing state records contained in a SignedEnvelope.
+const StateEnvelopeDomain = "libp2p-routing-state"
+
+// The type hint used to identify routing state records in a SignedEnvelope.
+// TODO: register multicodec
+var StateEnvelopePayloadType = []byte("/libp2p/routing-state-record")
+
+type SignedRoutingState struct {
+	// PeerID is the ID of the peer this record pertains to.
+	PeerID peer.ID
+
+	// Seq is an increment-only sequence counter used to order RoutingState records in time.
+	Seq uint64
+
+	// Addrs contains the public addresses of the peer this record pertains to.
+	Addrs []ma.Multiaddr
+
+	// Envelope contains the signature and serialized RoutingStateRecord protobuf.
+	// Although it uses a bit
+	Envelope *crypto.SignedEnvelope
+}
+
+// MakeSignedRoutingState returns a SignedRoutingState record containing the given multiaddrs,
+// signed with the given private key.
+// It generates a timestamp-based sequence number.
+func MakeSignedRoutingState(privKey crypto.PrivKey, addrs []ma.Multiaddr) (*SignedRoutingState, error) {
+	p, err := peer.IDFromPrivateKey(privKey)
+	if err != nil {
+		return nil, err
+	}
+	idBytes, err := p.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	seq := statelessSeqNo()
+	msg := pb.RoutingStateRecord{
+		PeerId:    idBytes,
+		Seq:       seq,
+		Addresses: addrsToProtobuf(addrs),
+	}
+	payload, err := proto.Marshal(&msg)
+	if err != nil {
+		return nil, err
+	}
+	envelope, err := crypto.MakeEnvelope(privKey, StateEnvelopeDomain, StateEnvelopePayloadType, payload)
+	if err != nil {
+		return nil, err
+	}
+	return &SignedRoutingState{
+		PeerID:   p,
+		Seq:      seq,
+		Addrs:    addrs,
+		Envelope: envelope,
+	}, nil
+}
+
+// UnmarshalSignedRoutingState accepts a serialized SignedEnvelope message containing
+// a RoutingStateRecord protobuf and returns a SignedRoutingState record.
+// Fails if the signature is invalid, if the envelope has an unexpected payload type,
+// if deserialization of the envelope or its inner payload fails.
+func UnmarshalSignedRoutingState(envelopeBytes []byte) (*SignedRoutingState, error) {
+	envelope, err := crypto.OpenEnvelope(envelopeBytes, StateEnvelopeDomain)
+	if err != nil {
+		return nil, err
+	}
+	return SignedRoutingStateFromEnvelope(envelope)
+}
+
+// SignedRoutingStateFromEnvelope accepts a SignedEnvelope struct containing
+// a RoutingStateRecord protobuf and returns a SignedRoutingState record.
+// Fails if the signature is invalid, if the envelope has an unexpected payload type,
+// or if deserialization of the envelope payload fails.
+func SignedRoutingStateFromEnvelope(envelope *crypto.SignedEnvelope) (*SignedRoutingState, error) {
+	if bytes.Compare(envelope.PayloadType, StateEnvelopePayloadType) != 0 {
```

Comment:
I realized this is probably too strict, since we might want to change the payload_type later if we register a new multicodec code for routing records. We could log a warning, but I don't think go-libp2p-core has a logging framework, so maybe it's better for this method to just ignore the payload type. Unmarshal will fail anyway if the payload is actually the wrong type.